ray-curator/ray_curator/stages/deduplication/fuzzy/minhash.py

```diff
@@ -0,0 +1,326 @@
+        self.char_ngrams = char_ngrams
+        self.num_hashes = num_hashes
+        self.seed = seed
+        self.use_64bit_hash = use_64bit_hash
+        self.read_format = read_format
+        self.read_kwargs = read_kwargs or {}
+        self.write_kwargs = write_kwargs or {}
+        self.pool = pool
+        # Initialize the minhash processor in setup
+        self.minhash_processor = None
+        self.id_generator = None
+
+    def setup(self, _worker_metadata: "WorkerMetadata | None" = None) -> None:
+        """Initialize the GPU MinHash processor and ID generator."""
+        # Initialize the ID generator (will be shared across workers)
+        self.id_generator = IdGenerator.options(
+            name=CURATOR_ID_GENERATOR_ACTOR_NAME, get_if_exists=True, lifetime="detached"
+        ).remote()
+
+        # Initialize the GPU minhash processor
+        self.minhash_processor = GPUMinHash(
+            seed=self.seed,
+            num_hashes=self.num_hashes,
+            char_ngrams=self.char_ngrams,
+            use_64bit_hash=self.use_64bit_hash,
+            pool=self.pool,
+        )
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        """Define input requirements."""
+        return (["data"], [])
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        """Define outputs - produces FileGroupTask with minhash files."""
+        return (["data"], [])
+
+    def process(self, task: FileGroupTask) -> FileGroupTask:
+        """
+        Process a group of files to compute minhashes.
+
+        Args:
+            task: FileGroupTask containing file paths to process
+
+        Returns:
+            FileGroupTask containing paths to minhash output files
+        """
+
+        if self.minhash_processor is None or self.id_generator is None:
+            msg = "MinHash processor or ID generator not initialized. Call setup() first."
+            raise RuntimeError(msg)
+
+        fs = get_fs(self.output_dir, self.write_kwargs.get("storage_options", {}))
+        output_file = fs.sep.join([self.output_dir, self._name, f"{task.task_id}.parquet"])
+
+        read_kwargs = self.read_kwargs.copy()
+        read_kwargs["storage_options"] = task._metadata.get("storage_options", {})
+
+        # Read input file based on format
+        if self.read_format == "jsonl":
+            df = self.read_jsonl(filepath=task.data, columns=[self.text_column], assign_id=True, **read_kwargs)
+        elif self.read_format == "parquet":
+            df = self.read_parquet(filepath=task.data, columns=[self.text_column], assign_id=True, **read_kwargs)
+        else:
+            msg = f"Unsupported read format: {self.read_format}"
+            raise ValueError(msg)
+
+        result_df = df[[CURATOR_DEDUP_ID_STR]]
+        result_df[self.minhash_column] = self.minhash_processor.compute_minhashes(df[self.text_column])
+
+        # Write output file
+        self.write_parquet(df=result_df, filepath=output_file, **self.write_kwargs)
```

Comment:
Same here, it is more beneficial/negligible impact to write here instead of as a separate stage?

Reply:
@sarahyurick the reason for this per my understanding is that df is cudf.DataFrame so passing it as DocumentBatch means it gets serialized to CPU and then transferred over network to the next stage which will write using CPU. Might as well write it here while we're at it